### ISSUE TITLE
[Reports] Customers report has a time shift between request and result

### ISSUE THREAD

**Issue by drummer83** [openfoodfoundation/openfoodnetwork#11607]
## Description

The Customers report has a time shift of 4 hours between the request and the results (on staging UK at least). Other report are not showing this behaviour. As a result it's difficult to find certain orders and the 'Last Completed Order Date' is always showing the day before the actual last completed order date.

## Expected Behavior

The report should display results according to the request (filter settings).

## Actual Behaviour

The report has a time shift between the results and the request (see screenshot for example).

## Steps to Reproduce

1. Pick an order from the orders list in the back office and note date and time of the order.
2. Run the Customers report and set the date and time filter to include only this order.
3. The order will not get returned as a result due to the time shift.
4. Widen the time settings of the filter until the order is found.
5. Compare the filter settings with the order date and time.

## Animated Gif/Screenshot

Here is an example:
- According to the DB the order was completed on 2023-09-30 09:59.
- The report filter has to be set to 2023-09-30 09:55 to 2023-09-30 10:00 to find the order.
- The order time is shown as 05:59 am in the back office.
- The 'Last completed order date' is shown as 2023-09-29.
![image](https://github.com/openfoodfoundation/wishlist/assets/1790176/d3b01d50-4c52-4193-9ec9-ba96d8127334)
![image](https://github.com/openfoodfoundation/wishlist/assets/1790176/35270f4c-afb7-48bb-8345-d7a7ce4a027a)

To find the same order in the Orders and Distributors report I have to set the filter to 5:55 to 06:00 (which matches the order time in the back office, but not in the DB).

## Workaround

None.

## Severity
bug-s3 or bug-s4

## Your Environment

* Version used: staged #11543 
* Browser name and version:
* Operating System and version (desktop or mobile):

**Comment by abdellani** [comment]
Hi  Konrad,

Thank  you for the detailed report :pray: 

I was able to reproduce the issue locally.
I created an order at:  "2023/10/02 18:18" ( My local timezone UTC+1, this means 17:18 on UTC+0)

I tried to filter orders between: "2023/10/03 18:00" and "2023/10/02 19:00" => no results.
I tried to filter orders between; "2023/10/03 17:00" and "2023/10/02 18:00" => I got the order.
As you can notice, there's a time shift of 1 hour (as I'm on UTC+1)
Germany is on (UTC+2), so you saw a time shift of 2 hours. 

We're filtering the orders as if the user is located on UTC+0.
To solve the issues, we'll need to send the Timezone along with the DateTime.

### Another bug
I noticed another bug.
The completion time in the report is converted to a date (from  AEDT) on the backend without considering the user's time zone. 

In other words, the report is showing the dates in Australia and not in the location of the user.

**Comment by abdellani** [comment]
I checked some other reports and I noticed similar problems.
I think that those bugs are not already reported. 
|Report|Description|
|-|-|
|Orders And Distributors|Date/time is filtered as if the user is on AEDT|
|Bulk Co-op *| same|
|Payments *|same|
|Order Cycle *|same|

I didn't check all the reports, but I think this is a general problem.

**Comment by dacook** [comment]
Something we could also consider is to show the instance timezone when displaying times. I think we could add that in the order confirmation email, admin orders screen.

Showing it for the the datepicker is probably not possible. Flatpickr doesn't support setting a timezone, we can't just hack it to append a timezone descriptor, because it depends on the selected date (eg in Melbourne, ` 2023-09-01 09:00` is `AEST`, but `2023-10-01 09:00` is `AEDT`)

**Comment by ccozkan** [comment]
Hi,
I am not sure if this is the required fix for this issue but I also experienced this issue when I was developing for PR https://github.com/openfoodfoundation/openfoodnetwork/pull/11619.
Just wanted to share that afai observed, I was able to solve the issue by changing 

https://github.com/openfoodfoundation/openfoodnetwork/blob/f93fcd347b073458909cee72a3908294e3f34195/lib/reporting/reports/customers/base.rb#L52-L55

to
```ruby
min = params.dig(:q, :completed_at_gt).presence&.in_time_zone
max = params.dig(:q, :completed_at_lt).presence&.in_time_zone
```

**Comment by abdellani** [comment]
Hi

I want to clarify the difference between the bug reported by @drummer83  and the bug I reported on my [comment](https://github.com/openfoodfoundation/openfoodnetwork/issues/11607#issuecomment-1743499797)

Let's take the example of the order I completed in my comment.
|system time (UTC+1)|in SQL (UTC time)| Instance time(AEDT)|
|-|-|-|
|`2023-10-02 18:17:39.290315`|`2023-10-02 17:17:39.290315`|`2023-10-03 04:17:39.290315`|

**Issue 1: customer report** 
I needed to set the time range in UTC to fetch the order
||Start|End|
|-|-|-|
|expected correct inputs (UTC+1)|`2023-10-02 18:00`|`2023-10-02 19:00`|
|current inputs that worked (UTC)|`2023-10-02 17:00`|`2023-10-02 18:00`|

**Issue 2: Orders And Distributes (+ others ) reports**
In this case, I needed to set the time range in AEDT

||Start|End|
|-|-|-|
|expected correct inputs (UTC+1)|`2023-10-02 18:00`|`2023-10-02 19:00`|
|current inputs that worked (AEDT)|`2023-10-03 04:00`|`2023-10-03 05:00`|

**Issue 3: Dates are calculated in the instance time zone**
In the customer order, the `LAST COMPLETED ORDER` is `2023-10-03` instead of `2023-10-02`.


I think, as @lin-d-hop shared, the first issue report by @drummer83  is the most important.
If the user is located in the same time zone as the server, issues 2 & 3 will not be noticed. It can be a problem if the user is located in a multi-time zone country.

@ccozkan 
I tested your solution and it fixes the issue. Nice job :+1: 


Side note:
`"2023-10-02 00:00".in_time_zone` is different from `DateTime.parse("2023-10-02 00:00").in_time_zone`
I wanted to share this as I wasted some time trying `DateTime`.
